nucliadb_vectors/src/data_point/ops_hnsw.rs

```diff
@@ -129,37 +128,54 @@ impl<'a, DR: DataRetriever> HnswOps<'a, DR> {
         let picked_level = -sample.ln() * params::level_factor();
         picked_level.round() as usize
     }
-    fn closest_up_node<L: Layer>(
+    fn closest_up_nodes<L: Layer>(
         &'a self,
-        x: Address,
+        entry_points: Vec<Address>,
         query: Address,
         layer: L,
-        filter: NodeFilter<'a, DR>,
-    ) -> Option<(Address, f32)> {
+        number_of_results: usize,
+        mut filter: NodeFilter<'a, DR>,
+    ) -> Vec<(Address, f32)> {
         // We just need to perform BFS, the replacement is the closest node to the actual
         // best solution. This algorithm takes a lazy approach to computing the similarity of
         // candidates.
-        let mut visited_nodes = HashSet::new();
-        let mut candidates = VecDeque::from([x]);
+
+        let mut results = Vec::new();
+        let mut visited_nodes: HashSet<_> = HashSet::from_iter(entry_points.iter().copied());
+        let mut candidates = VecDeque::from(entry_points);
+
         loop {
-            let best_so_far = candidates.pop_front();
+            if results.len() == number_of_results {
+                return results;
+            }
 
-            match best_so_far.map(|n| (n, self.similarity(n, query))) {
-                None => break None,
-                Some((_, score)) if score < self.tracker.min_score() => break None,
-                Some((n, score)) if filter.is_valid(n, score) && filter.passes_formula(n) => break Some((n, score)),
-                Some((down, _)) => {
-                    let mut sorted_out: Vec<_> = layer.get_out_edges(down).collect();
-                    sorted_out.sort_by(|a, b| b.1.total_cmp(&a.1));
-                    sorted_out.into_iter().for_each(|(new_candidate, _)| {
-                        if !visited_nodes.contains(&new_candidate) {
-                            candidates.push_back(new_candidate);
-                            visited_nodes.insert(new_candidate);
-                        }
-                    });
-                }
+            let Some(candidate) = candidates.pop_front() else {
+                break;
+            };
+
+            let candidate_similarity = self.similarity(query, candidate);
+
+            if candidate_similarity < self.tracker.min_score() {
+                break;
             }
+
+            if filter.is_valid(candidate, candidate_similarity) && filter.passes_formula(candidate) {
+                let candidate_vector = self.tracker.get_vector(candidate);
+                filter.vec_counter.add(candidate_vector);
+                results.push((candidate, candidate_similarity));
+            }
```

Comment:
You might want to move the `if results.len() == number_of_results` to here? No need to expand the last nodes candidates

Reply:
Done! Thanks!